### PR TITLE
Millhone CLI: `analyze`

### PR DESCRIPTION
**Pull request by jssblck**
# Overview

Adds the `analyze` subcommand to Millhone, which fingerprints a local project and reports snippet matches.

Currently, it just writes these matches to the logger. Also, note that similar to the ingestion, matching currently runs single threaded.

## Acceptance criteria

Users can get raw matches.

## Testing plan

Automated tests cover the fiddly bits. I manually ran a smoke test:

In one terminal, run Millhone. I ran it with the in-memory DB for convenience:
```shell
foundation/millhone on ⑆ millhone/nits [$] is 📦 v0.2.0 via 🦀 v1.72.0
❯ cargo run -- --db-in-memory
    Finished dev [unoptimized + debuginfo] target(s) in 0.21s
     Running `/Users/jessica/projects/foundation/target/debug/millhone --db-in-memory`
  2023-09-20T01:40:36.556740Z  INFO millhone: creating server, options: Application { port: 3000, tracing: TracingConfig { trace_level: Info, trace_spans: Off, trace_format: Text, trace_colors: Auto }, app_db: AppDbConfig { app_db_host: NonEmptyString("localhost"), app_db_port: 5432, app_db_name: NonEmptyString("millhone"), app_db_user: NonEmptyString("fossa"), app_db_password: Secret([REDACTED alloc::string::String]), app_db_max_pool: 10, app_db_min_pool: 10, app_db_connection_timeout_seconds: 30 }, corpus_db: CorpusDbConfig { corpus_db_nodes: ["localhost:9042"], corpus_db_keyspace: "millhone", corpus_db_username: None, corpus_db_password: None }, db_in_memory: true }
    at millhone/src/main.rs:198

  2023-09-20T01:40:36.567071Z  INFO millhone: booting server, addr: 0.0.0.0:3000
    at millhone/src/main.rs:202
```

I then ingested the `cpp-vsi-demo` project:
```shell
fossa-cli/extlib/millhone on ⑆ millhone-cli/analyze [$!?] is 📦 v0.1.0 via 🦀 v1.72.0
❯ cargo run -- ingest --direct-endpoint http://localhost:3000 --api-key-id 0ee7fb68-d368-4ee1-8a81-9bb77a722067 --api-secret cf7a2814-f9f6-4d15-9828-5bd750ee4246 --locator 'git+github.com/fossas/cpp-vsi-demo-src$v0.1.0' ~/projects/cpp-vsi-demo/example-internal-project --trace-level off
    Finished dev [unoptimized + debuginfo] target(s) in 0.07s
     Running `/Users/jessica/projects/fossa-cli/target/debug/millhone ingest --direct-endpoint 'http://localhost:3000' --api-key-id 0ee7fb68-d368-4ee1-8a81-9bb77a722067 --api-secret cf7a2814-f9f6-4d15-9828-5bd750ee4246 --locator 'git+github.com/fossas/cpp-vsi-demo-src$v0.1.0' /Users/jessica/projects/cpp-vsi-demo/example-internal-project --trace-level off`
```

I then ran this new subcommand and observed matches as expected:
```shell
fossa-cli/extlib/millhone on ⑆ millhone-cli/analyze [$!?] is 📦 v0.1.0 via 🦀 v1.72.0
❯ cargo run -- analyze --direct-endpoint http://localhost:3000 --api-key-id 0ee7fb68-d368-4ee1-8a81-9bb77a722067 --api-secret cf7a2814-f9f6-4d15-9828-5bd750ee4246 ~/projects/cpp-vsi-demo/example-internal-project/src
    Finished dev [unoptimized + debuginfo] target(s) in 0.07s
     Running `/Users/jessica/projects/fossa-cli/target/debug/millhone analyze --direct-endpoint 'http://localhost:3000' --api-key-id 0ee7fb68-d368-4ee1-8a81-9bb77a722067 --api-secret cf7a2814-f9f6-4d15-9828-5bd750ee4246 /Users/jessica/projects/cpp-vsi-demo/example-internal-project/src`
2023-09-20T01:41:12.674004Z  INFO main{target=/Users/jessica/projects/cpp-vsi-demo/example-internal-project/src}: millhone::cmd::analyze: Analyzing local snippet matches api_key_id=0ee7fb68-d368-4ee1-8a81-9bb77a722067
2023-09-20T01:41:12.811752Z  INFO main{target=/Users/jessica/projects/cpp-vsi-demo/example-internal-project/src}: millhone::cmd::analyze: analyze path=/Users/jessica/projects/cpp-vsi-demo/example-internal-project/src/main.c
2023-09-20T01:41:12.816400Z  INFO main{target=/Users/jessica/projects/cpp-vsi-demo/example-internal-project/src}: millhone::cmd::analyze: extracted snippets snippet_count=1
2023-09-20T01:41:12.819300Z  INFO main{target=/Users/jessica/projects/cpp-vsi-demo/example-internal-project/src}:lookup_snippets{fingerprint=RjpGXHcd2yCz04Q+BXBT3w65hnLBbXZE+zcmBW4OPw0=}:run{url="http://localhost:3000/api/v1/lookup/RjpGXHcd2yCz04Q-BXBT3w65hnLBbXZE-zcmBW4OPw0="}: millhone::api::v1::lookup: fetched matching snippets retry=1
2023-09-20T01:41:12.824651Z  INFO main{target=/Users/jessica/projects/cpp-vsi-demo/example-internal-project/src}: millhone::cmd::analyze: matching snippet: ApiSnippet {
    snippet: Snippet {
        fingerprint: Fingerprint(
            "RjpGXHcd2yCz04Q+BXBT3w65hnLBbXZE+zcmBW4OPw0=",
        ),
        target: "function",
        kind: "full",
        method: "raw",
        file_path: "src/main.c",
        byte_start: 21,
        byte_end: 73,
        line_start: 4,
        line_end: 7,
        col_start: 1,
        col_end: 2,
        language: "c99_tc3/static",
    },
    locator: Locator {
        fetcher: Git,
        org_id: None,
        project: "github.com/fossas/cpp-vsi-demo-src",
        revision: Some(
            "v0.1.0",
        ),
    },
    ingest_id: "675a0c18-82a8-4acf-950f-0b4a7927b439",
} fingerprint=RjpGXHcd2yCz04Q+BXBT3w65hnLBbXZE+zcmBW4OPw0=
2023-09-20T01:41:12.824876Z  INFO main{target=/Users/jessica/projects/cpp-vsi-demo/example-internal-project/src}: millhone::cmd::analyze: analyze path=/Users/jessica/projects/cpp-vsi-demo/example-internal-project/src/main.h
2023-09-20T01:41:12.824952Z  INFO main{target=/Users/jessica/projects/cpp-vsi-demo/example-internal-project/src}: millhone::cmd::analyze: extracted snippets snippet_count=0
2023-09-20T01:41:12.824987Z  INFO main{target=/Users/jessica/projects/cpp-vsi-demo/example-internal-project/src}: millhone::cmd::analyze: Finished extracting snippets total_count_entries=3 total_count_snippets=1 total_count_files=2
```

This looks reasonable, so I'm happy with this result.

## Risks

None

## Metrics

None

## References

Implements https://fossa.atlassian.net/browse/ANE-1114

## Checklist

- [x] I added tests for this PR's change (or explained in the PR description why tests don't make sense).

### CODE REVIEW

**Comment by jssblck** [comment]
@spatten, tagged you for review on this because it overlaps fairly heavily with #1278.

**Review by spatten — APPROVED** [review]
Looks good to me!

File: extlib/millhone/src/extract.rs [hunk 1]
```diff
@@ -0,0 +1,557 @@
...
+    Unsupported,
+
+    /// The language is supported.
+    Supported(Language),
+}
+
+impl Support {
+    /// Determine whether a file is supported, and its language, based on its extension.
+    pub fn by_ext(path: &Path) -> Self {
+        path.extension()
```
[the remark this answers is not in the document]
Comment (jssblck) [reply extlib/millhone/src/extract.rs:428]: I just wanted to be maximally correct; in the future I could see us doing something like "well if it's unknown, try to resolve it some other way (like my file content)".

File: extlib/millhone/src/extract.rs [hunk 2]
```diff
@@ -0,0 +1,557 @@
...
+        let decoded = BASE64_STANDARD
+            .decode(raw)
+            .map_err(serde::de::Error::custom)?;
+        Self(decoded).pipe(Ok)
+    }
+}
+
+/// The location, using textual "line, column" indicators, for a snippet.
+///
+/// Line end is inclusive: if a snippet starts on line 1 and ends before
```
[the remark this answers is not in the document]
Comment (jssblck) [reply extlib/millhone/src/extract.rs:340]: Addressed in #1283: https://github.com/fossas/fossa-cli/pull/1283/commits/15397a99d799d9d1cf46e6edbfe791d6683a9e89

**Pull request merged**